src/Features/LanguageServer/Protocol/Handler/Configuration/DidChangeConfigurationNotificationHandler.cs

```diff
@@ -0,0 +1,98 @@
+        private readonly Guid _registrationId;
+
+        public DidChangeConfigurationNotificationHandler(ILspLogger logger, IGlobalOptionService globalOptionService, IClientLanguageServerManager clientLanguageServerManager)
+        {
+            _lspLogger = logger;
+            _globalOptionService = globalOptionService;
+            _clientLanguageServerManager = clientLanguageServerManager;
+            _registrationId = Guid.NewGuid();
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public bool RequiresLSPSolution => true;
+
+        [LanguageServerEndpoint(Methods.WorkspaceDidChangeConfigurationName)]
+        public Task HandleNotificationAsync(DidChangeConfigurationParams request, RequestContext requestContext, CancellationToken cancellationToken)
+            => RefreshOptionsAsync(cancellationToken);
+
+        private async Task RefreshOptionsAsync(CancellationToken cancellationToken)
+        {
+            var globalConfigurationItems = s_supportedGlobalOptions.SelectAsArray(option => new ConfigurationItem() { ScopeUri = null, Section = option.Definition.ConfigName });
+            var perLanguageConfigurationItems = s_supportedPerLanguageOptions.SelectAsArray(option => new ConfigurationItem() { ScopeUri = null, Section = option.Definition.ConfigName });
+            var allConfigurationItems = globalConfigurationItems.Concat(perLanguageConfigurationItems);
+
+            var configurationsFromClient = await GetConfigurationsAsync(allConfigurationItems, cancellationToken).ConfigureAwait(false);
+            for (var i = 0; i < globalConfigurationItems.Length; i++)
+            {
+                var globalOptions = s_supportedGlobalOptions[i];
+                var configurationValue = configurationsFromClient[i];
+                globalOptions.WriteToGlobalOptionService(_globalOptionService, configurationValue);
+            }
+
+            for (var i = 0; i < perLanguageConfigurationItems.Length; i++)
+            {
+                var perLanguageOptions = s_supportedPerLanguageOptions[i];
+                var configurationValue = configurationsFromClient[i + globalConfigurationItems.Length];
+                perLanguageOptions.WriteToGlobalOptionService(_globalOptionService, LanguageNames.CSharp, configurationValue);
+            }
+        }
+
+        private async Task<ImmutableArray<string>> GetConfigurationsAsync(ImmutableArray<ConfigurationItem> configurationItems, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var configurationParams = new ConfigurationParams() { Items = configurationItems.AsArray() };
+                var options = await _clientLanguageServerManager.SendRequestAsync<ConfigurationParams, JArray>(
+                    Methods.WorkspaceConfigurationName, configurationParams, cancellationToken).ConfigureAwait(false);
+
+                if (options == null)
```

Comment:
if all of these are truly unexpected errors that we never should hit, just throw - the queue should log exceptions.

But be careful here - as any exception in a mutating handler will tear down the entire server.